libnmstate/state.py

```diff
@@ -32,6 +32,68 @@
 from libnmstate.schema import Interface
 from libnmstate.schema import InterfaceState
 from libnmstate.schema import InterfaceType
+from libnmstate.schema import Route
+
+
+class RouteEntry(object):
+
+    # Mapping dict key name to RouteEntry property name
+    KEY_PROPETY_MAP = {
+        Route.STATE: Route.STATE.replace('-', '_'),
+        Route.TABLE_ID: Route.TABLE_ID.replace('-', '_'),
+        Route.METRIC: Route.METRIC.replace('-', '_'),
+        Route.NEXT_HOP_ADDRESS: Route.NEXT_HOP_ADDRESS.replace('-', '_'),
+        Route.DESTINATION: Route.DESTINATION.replace('-', '_'),
+        Route.NEXT_HOP_INTERFACE: Route.NEXT_HOP_INTERFACE.replace('-', '_'),
+    }
+
+    def __init__(self, route):
+        self.table_id = route.get(Route.TABLE_ID)
+        self.state = route.get(Route.STATE)
+        self.metric = route.get(Route.METRIC)
+        self.destination = route.get(Route.DESTINATION)
+        self.next_hop_address = route.get(Route.NEXT_HOP_ADDRESS)
+        self.next_hop_interface = route.get(Route.NEXT_HOP_INTERFACE)
+        self.complement_defaults()
+
+    def complement_defaults(self):
+        if self.state != Route.STATE_ABSENT:
+            if self.table_id is None:
+                self.table_id = Route.USE_DEFAULT_ROUTE_TABLE
+            if self.metric is None:
+                self.metric = Route.USE_DEFAULT_METRIC
+            if self.next_hop_address is None:
+                self.next_hop_address = ''
+
+    def __hash__(self):
+        return hash(
+            tuple(getattr(self, prop)
+                  for prop in six.viewvalues(RouteEntry.KEY_PROPETY_MAP)))
+
+    def to_dict(self):
+        route_entry = {}
+        for key, prop in six.viewitems(RouteEntry.KEY_PROPETY_MAP):
+            if getattr(self, prop) is not None:
+                route_entry[key] = getattr(self, prop)
+        return route_entry
+
+    def __eq__(self, other):
+        return self is other or self.to_dict() == other.to_dict()
+
+    def can_remove(self, other):
+        """
+        Return True if other route should be removed by self absent route.
+        """
+        if self.state != Route.STATE_ABSENT:
+            return False
+
+        for key, prop in six.viewitems(RouteEntry.KEY_PROPETY_MAP):
+            if key == Route.STATE:
```

Comment:
what if the other route entry has the value `absent` for the `state` property?

Reply:
Good point. In that case a absent cannot remove another absent.

Reply:
Test case added also.